data.go

```diff
@@ -85,86 +64,289 @@ func setDB(db apid.DB) {
 
-	stmt, err := txn.Prepare("DELETE FROM DATA_SCOPE WHERE id=$1 and apid_cluster_id=$2")
-	if err != nil {
-		log.Errorf("update DATA_SCOPE failed: %v", err)
-		return err
+	for _, pk := range pkeys {
+		wherePlaceholders = append(wherePlaceholders, fmt.Sprintf("%s=$%v", pk, i))
+		i++
 	}
-	defer stmt.Close()
 
-	_, err = stmt.Exec(ds.ID, ds.ClusterID)
+	sql := "DELETE FROM " + normalizedTableName
+	sql += " WHERE "
+	sql += strings.Join(wherePlaceholders, " AND ")
+
+	return sql
+
+}
+
+func update(tableName string, oldRows, newRows []common.Row, txn *sql.Tx) bool {
+	pkeys, err := getPkeysForTable(tableName)
+	if len(pkeys) == 0 || err != nil {
+		log.Errorf("UPDATE No primary keys found for table.", tableName)
+		return false
+	} else {
+		if len(oldRows) == 0 || len(newRows) == 0 {
+			return false
+		}
+
+		var orderedColumns []string
 
+		//extract sorted orderedColumns
+		for columnName := range newRows[0] {
+			orderedColumns = append(orderedColumns, columnName)
+		}
+		sort.Strings(orderedColumns)
+
+		//build update statement, use arbitrary row as template
+		sql := buildUpdateSql(tableName, orderedColumns, newRows[0], pkeys)
+		prep, err := txn.Prepare(sql)
+		if err != nil {
+			log.Errorf("UPDATE Fail to prep statement [%s] error=[%v]", sql, err)
+			return false
+		}
+		defer prep.Close()
+
+		for i, row := range newRows {
+			var values []interface{}
+
+			for _, columnName := range orderedColumns {
+				//use Value so that stmt exec does not complain about common.ColumnVal being a struct
+				//TODO will need to convert the Value (which is a string) to the appropriate field, using type for mapping
+				//TODO right now this will only work when the column type is a string
+				if row[columnName] != nil {
+					values = append(values, row[columnName].Value)
+				} else {
+					values = append(values, nil)
+				}
+			}
+
+			//add values for where clause, use PKs of old row
+			for _, pk := range pkeys {
+				if oldRows[i][pk] != nil {
+					values = append(values, oldRows[i][pk].Value)
+				} else {
+					values = append(values, nil)
+				}
+
+			}
+
+			//create prepared statement from existing template statement
+			res, err := txn.Stmt(prep).Exec(values...)
+
+			if err != nil {
+				log.Errorf("UPDATE Fail [%s] values=%v error=[%v]", sql, values, err)
+				return false
+			} else {
+				numRowsAffected, err := res.RowsAffected()
+				if err != nil {
+					log.Errorf("UPDATE Fail [%s] values=%v error=[%v]", sql, values, err)
+					return false
+				}
+				//delete this once we figure out why tests are failing/not updating
+				log.Infof("NUM ROWS AFFECTED BY UPDATE: %d", numRowsAffected)
+				log.Debugf("UPDATE Success [%s] values=%v", sql, values)
+			}
+		}
+
+		return true
+	}
+}
+
+func buildUpdateSql(tableName string, orderedColumns []string, row common.Row, pkeys []string) string {
+	if row == nil {
+		return ""
+	}
+	normalizedTableName := normalizeTableName(tableName)
+
+	var setPlaceholders, wherePlaceholders []string
```

Comment:
actually this one I'll leave, its messier to preset the length because then you can't `append` and you have to create a new index variable, etc